[WIP] Reorganize and update documentation

OriolAbril opened the pull request:
will close #4044 and will close #4894

+ [ ] what are the (breaking) changes that this PR makes? This will break a lot of links as once merged, `latest` (and after release `stable`) will stop having notebooks build within the docs, many pages will be renamed or moved...
+ [ ] important background, or details about the implementation. This is key GSoD work which also build on top of recent improvements around notebooks and versioned docs.
+ [ ] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md. This will change the face of the website considerably, so it would be great to sync with major 4.0 release and advertize improved docs with the release

OriolAbril committed:
Getting docs ready for v4

example doc restructure

working and hopefully illustrative doc skeleton

update requirements

WIP
Change log:
- Upload new PyMC logo
- Updated conf.py to pymc instead of pymc3
- Start configuring and populating index page
- Start configuring and populating learning page
- Start configuring and populating api page
- Updated API to follow the docs/Architecture.png layout
- Start configuring and populating learning page
- Fix typo in filename of "index legacy"

add community page, other minor changes

Continue restructuring the API
Delete conflicting files

example doc restructure

update requirements

update requirements and gitignore

some fixes

add changes introduced by pre-commit checks

fix community page title

update api for distributions

update gaussian process api

fix end of file

more changes

sponsor logos

sponsor logos cleanup, other changes

changes to satisfy pre-commit checks

requirements changes

minor changes

Add testimonials to docs homepage

remove sphinx panels and other changes

center buttons

add explanation of MAP to glassor

minor change

fix typo

minor changes

put content in videos and podcasts section. Change formatting of testimonials (width of 2 cards for better formating in narrow screens)

add code example to front page

revise some definitions (#5126)

minor changes

minor changes

WIP: organizing notebooks into learning section

martinacantaro commented:
Still working on it, many TO-DOs and things to be fixed

OriolAbril commented on docs/source/community.md:
```diff
@@ -0,0 +1,56 @@
+
+## Code of Conduct
+
+In the interest of fostering an open and welcoming environment, we as contributors and maintainers pledge to making participation in our project and our community a harassment-free experience for everyone, regardless of age, body size, disability, ethnicity, sex characteristics, gender identity and expression, level of experience, education, socio-economic status, nationality, personal appearance ...
```
I think this language is too strong (not completely sure about all meanings of pledge). We can say we are commited or dedicated to an inclusive community, but we can't really guarantee (like any other open source project or open space).

some examples that could serve as inspiration: https://numfocus.org/code-of-conduct or https://jupytercon.com/codeofconduct/

martinacantaro replied:
It's actually a placeholder copypasta :P will come up with something better

OriolAbril commented on docs/source/community.md:
```diff
@@ -0,0 +1,56 @@
+
+:::{raw} html
+
+  <a class="twitter-timeline" data-width="300" data-height="500" data-dnt="true" href="https://twitter.com/pymc_devs?ref_src=twsrc%5Etfw">Tweets by pymc_devs</a> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
```
Looking at the preview, this would probably look better on the sidebar. I think it won't be difficult to implement.

OriolAbril commented on docs/source/community.md:
```diff
@@ -0,0 +1,56 @@
+
+[Discourse](https://discourse.pymc.io/)
+:::
+::::
```
and this should probably go in the actual footer so it automatically appears in all pages.

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,104 @@
+:::{card} 
```
general comment on the page, we now use a custom template for the home page in order to use all the available width, including the spaces generally reserved for the sidebars. I would therefore strongly advise against having text that covers this full span. Sites always shrink the readable width so it's easier to read

martinacantaro replied:
Yes, that was a next step! since cards don't allow for width I wanted to look into two options (suggestions welcome!):
- Look into applying CSS to shrink the cards down
- Put them inside a grid with just 1 column and shrinking the column

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,104 @@
+:width: 50%
+---
+scale: 80
+---
```
you could use `:scale: 80` too, and I'd recommend either doing that or changing to the `---` fences syntax for everything. right now you are combining the two possible syntaxes for specifying directive arguments in myst: https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#parameterizing-directives

martinacantaro replied:
Cool, I'll remove all instances of `---`, I think were already there since I used `:` for everything

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,104 @@
+:shadow:
+
+Get started!
+:::
```
For these buttons we have to find the way of getting them centered I think. The quick and hacky way is using a 3 cols grid and placing the button at the central one with the :expand: flag, but we could also use the `div` directive (sphinx built-in one) with custom classes for that. sphinx design for example or raw bootstrap should have custom classes that allow this

ref https://sphinx-design.readthedocs.io/en/sbt-theme/css_classes.html

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,104 @@
+
+::::{grid} 1 3 3 3
+:::{grid-item-card}
+TODO add sponsor logo
```
reminder to myself to add also pymcon sponsors here as promised.

martinacantaro replied:
Adding the two current sponsors, numfocus and PyMC Labs

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,104 @@
+
+# Citing PyMC
+
+Use this to cite the library:
```
cc @michaelosthege we should probably add a mention to zenodo somewhere I think

michaelosthege replied:
We mention it on the Github landing page IIRC

martinacantaro replied:
Zenodo is only mentioned like this in the readme: `DOIs for specific versions are shown on Zenodo and under Releases`

OriolAbril commented on docs/source/installation.md:
```diff
@@ -0,0 +1,97 @@
+:::
+::::
+
+## Aesara Installation
```
I think those should only be links to the actual installation guides (which we could write/improve ourselves). Best case scenario they will be duplicated and if there are version dependent changes the instructions here won't be able to take that into account. Worst case scenario, they end up being outdated but as they are here users read these instead of the ones in the actual library and end up being more confused than what they would have been with no intallation guide nor link whatsoever. 

We are already in quite the worse case scenario ourselves, so this is something to be very careful about. Currently the right pymc installation guides are on wiki pages but the docs also have installation guides (outdated) and no link to the wiki pages

martinacantaro replied:
Done. Docs only redirect to the github wiki now. But the github wiki is also very outdated, so we need to:
- Create a v4 install guide for macos, windows and linux in the wiki, also linked to the docs website
- Keep the PyMC3 install guide but make sure they're clearly differentiated both in the wiki and in the website 
- Add info for users dealing with the library renaming, which is present in this doc https://hackmd.io/oB_ahfcIRKegO6hR0OW8eg

OriolAbril reviewed (commented):
quick review, tried to focus on fail fast comments on things that I _think_ you are probably following the wrong path or lack some context. Hopefully this will prevent you from spending too much time on things that then need to be changed

twiecki commented:
Where do the conflicts come from?

Also, is there a place to see what it'd look like?

OriolAbril commented:
> Where do the conflicts come from?

I started setting this up 1-2 months ago so it's probably renaming changes.

> Also, is there a place to see what it'd look like?

Yes, all PRs have a preview of how will their docs look like, you have to search for the readthedocs check, wait until it finishes and click "Details"

michaelosthege commented:
While you're at it, can you fix the coverage badge link on the docs landing page?

aloctavodia commented on docs/source/api.rst:
```diff
@@ -13,20 +8,14 @@ API Reference
+   api/smc
+   api/step_methods
+   api/inference
+   api/plots
```
Now we have a BART module #5058

martinacantaro replied:
Cool, I was going to ask more questions about the api resources, including BART, because we've been talking a bit with @OriolAbril about what needs to be displayed for users, what needs to be displayed for devs and what doesn't need to be displayed at all.
We were basing the current API tree on the [Architecture diagram](https://github.com/pymc-devs/pymc/blob/main/ARCHITECTURE.md), but that is probably out of date (as evidenced by it not including BART). 
What other things should be included/excluded? 
For example, does logprob submodule in distributions need to be added?

martinacantaro replied:
Adding comment by @Sayam753 :

Yes. At least, `logp` and `logcdf` functions should be exposed

OriolAbril commented on conda-envs/environment-dev-py37.yml:
```diff
@@ -22,9 +22,8 @@ dependencies:
-- sphinx-autobuild>=0.7
+- sphinx-copybutton
 - sphinx-notfound-page
 - sphinx-panels
```
remember to eventually remove sphinx-panels and use only sphinx-design

OriolAbril commented on docs/source/learn/notebooks.md:
```diff
@@ -1,3 +1,31 @@
+
+- https://docs.pymc.io/en/stable/pymc-examples/examples/diagnostics_and_criticism/model_comparison.html
+  - https://github.com/pymc-devs/pymc-examples/blob/main/examples/diagnostics_and_criticism/model_comparison.ipynb
+  TODO: Add reference to ArviZ docs since it was taken from there
```
I don't think it was taken from ArviZ, probably from somewhere else

OriolAbril commented on docs/source/learn/notebooks.md:
```diff
@@ -1,3 +1,31 @@
+  - https://github.com/pymc-devs/pymc/blob/1048b69ac1410f6dec64eb1f26778d89b28a62c4/docs/source/Probability_Distributions.rst
+
+- https://docs.pymc.io/en/stable/PyMC3_and_Theano.html
+  update needed: from "pymc and theano" to "pymc and aesara" tutorial (not just auto-renaming from theano to aesara but showing correct use). Very related to migration guide as well, though the migration guide already covers it.
```
do we have a migration guide?

OriolAbril committed:
add style guides

martinacantaro committed:
remove multilevel modeling

martinacantaro committed:
formatting the learning section

martinacantaro committed:
add anchors to notebooks

martinacantaro committed:
add anchors to all the main pages for easy intra-doc linking

OriolAbril commented on docs/source/Advanced_usage_of_Aesara_in_PyMC.rst:
```diff
@@ -1,5 +1,5 @@
 :orphan:
-
+(Advanced_usage_of_Aesara_in_PyMC)
```
```suggestion
(Advanced_usage_of_Aesara_in_PyMC)=
```

las anchors en myst deben terminar con el igual. Yo también trataria de usar ids mas cortos para las anchors, por ejemplo `advanced_aesara` (no creo que tengamos paginas de aesara usage fuera de pymc, y si las tuvieramos esas podrian tener id mas largo y especifico)

martinacantaro replied:
Uy, eso fue un error de tipeo.
Estaba en la duda con respecto a usar nombres largos o cortos. Opté por `título del notebook` =` título del archivo` = `nombre del anchor`, para no tener que memorizar varios nombres para una misma cosa (y no los vamos a tener que escribir tantas veces que amerite abreviarlo), que opinas?

martinacantaro committed:
Changelog
- Add information for institutions in the Contributing section
- Center navbar
- Split the About section into a main page with multiple subpages
- Remove old reference to "developers" in the index toctree

martinacantaro committed:
fix anchor for Advanced_usage_of_Aesara_in_PyMC.rst

Co-authored-by: Oriol Abril-Pla <oriol.abril.pla@gmail.com>

OriolAbril commented on docs/source/about/citing_pymc.md:
```diff
@@ -0,0 +1,18 @@
+
+The BibTeX entry is:
+
+.. code-block:: none
```
This is rst directive, but the file is markdown. It should be

````suggestion
```bibtex
````

Plus no indents and closing backtick fence

OriolAbril commented on docs/source/about/citing_pymc.md:
```diff
@@ -0,0 +1,18 @@
+
+.. code-block:: none
+
+    @article{pymc,
```
Looks like doi is missing here comparing with citation above

OriolAbril commented on docs/source/about/citing_pymc.md:
```diff
@@ -0,0 +1,18 @@
+      pages={e55},
+      year={2016},
+      publisher={PeerJ Inc.}
+    }
```
Add todo or note on clickup (might be done already) about adding zenodo here,  the readme on pyms repo has the right citation info

OriolAbril commented on docs/source/about/pymc_for_enterprise.md:
```diff
@@ -0,0 +1,62 @@
+let's face it, amazing!
+
+
+TODO fix code
```
More than fix ckde, use sphinx design buttons. Otherwise the next time bootstrap is updated the buttons will break again

OriolAbril commented on docs/source/conf.py:
```diff
@@ -46,11 +45,11 @@
     "sphinx_design",
-    "sphinx_panels",
     "notfound.extension",
+    "sphinx_design",
```
Sphinx desing looks duplicated

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+After the notebook content finishes, there should be an `## Authors` section with bullet points
+to provide attribution to the people who contributed to the the general pattern should be:
+
+```
````
````suggestion
```markdown
````

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+  It will therefore follow a different structure than the example above
+  in order to include a link or reference to the original source:
+
+  ```
````
````suggestion
  ```markdown
````

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+
+some examples:
+
+```
````
````suggestion
```markdown
````

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+
+and
+
+```
````
````suggestion
```markdown
````

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+and
+
+```
+* Adapted from chapter 5 of Bayesian Data Analysis 3rd Edition {cite:p}gelman2013bayesian
````
```suggestion
* Adapted from chapter 5 of Bayesian Data Analysis 3rd Edition {cite:p}`gelman2013bayesian`
```

OriolAbril commented on docs/source/contributing/jupyter_style.md:
````diff
@@ -0,0 +1,221 @@
+%watermark -n -u -v -iv -w -p theano,xarray
+```
+
+This last code cell should be preceded by a markdown cell with the `## Watermark` title only so it appears in the table of contents.
````
```suggestion
This second to last code cell should be preceded by a markdown cell with the `## Watermark` title only so it appears in the table of contents.
```

OriolAbril commented on docs/source/index.md:
```diff
@@ -0,0 +1,197 @@
+
+::::{card-carousel} 2
+
+:::{card}
```
Todos/considerations for future PRs. Make the whole card clickable and consider using epigraph formatting https://jupyterbook.org/content/content-blocks.html#epigraphs

martinacantaro committed:
redoing the community page after it got lost in a force push

martinacantaro committed:
minor changes

martinacantaro committed:
update code example

martinacantaro committed:
redo community page

OriolAbril committed:
try twitter feed on sidebar

OriolAbril committed:
restructure contributing docs

OriolAbril committed:
fix requirements

OriolAbril committed:
Apply suggestions from code review
